decompiler/analysis/insert_lets.cpp

```diff
@@ -699,6 +699,56 @@ FormElement* rewrite_empty_let(LetElement* in, const Env&, FormPool&) {
   return in->entries().at(0).src->try_as_single_element();
 }
 
+FormElement* rewrite_set_let(LetElement* in, const Env& env, FormPool& pool) {
+  /*
+   * (let ((dest-var src))
+   *   (set! something dest-var)
+   *   dest-var
+   *   )
+   * to:
+   * (set! something src)
```

Comment:
I think this change is unsafe to make - for example:
```
(let ((temp (foo)))
  (set! (-> (bar) field) temp)
  )
```
will call `foo` before `bar`, but

```
(set! (-> (bar) field) (foo))
```
will call `bar` before `foo`.

Which is why I think this detection has to happen before expression building.
If we do it here, we also lose the ability to detect things like `*!`

Reply:
I was wondering that too, but it looks like `set!` is an exception to the left-to-right eval order. The source is eval'd before the destination.
![image](https://user-images.githubusercontent.com/7569514/188774994-6aa7d9db-ef2d-4de3-b0d8-0402d9d7a67a.png)

Reply:
Math macros like `*!` also seemed unaffected, looking at the refs.

Reply:
Yeah it looks like you're right - both the compiler and decompiler agree about this, and I made a comment about it being flipped here
![image](https://user-images.githubusercontent.com/48171810/188781066-f240a40f-f56c-4eec-a52a-aeeac51727d1.png)

I think that means this is safe, but I do want to think about it a little more.

The `*!` thing would only apply in cases where we don't detect `*!` currently - if you had something that was originally `(*! foo bar)`, but it got split up due to this set/let issue, then this fix will give us `(set! foo (* foo bar))`. Which I wouldn't worry about too much.

A few other things I want to think about
- `uint128`/`vf` stuff : seems like it doesn't get applied for vf stuff because that doesn't become a `set`
- dropping casts for `float`/`int` convers: I think it won't but I want to review some more
- dropping casts for `(the-as ` - not sure, I kinda think this will be wrong
- bitfield stuff: the compile makes `set!` on a bitfield return `none`, but we could change this